### ISSUE TITLE
Code editor quality of life

### ISSUE THREAD

**Issue by drasmuss** [nengo/nengo_gui#427]
It'd be nice if the code editor would auto-indent nicely on line breaks e.g.
```python
func(a,
     b)
```

It'd be nice if the code editor used spaces instead of tabs on indent

It'd be nice if ctrl-s saved the file when the editor has the focus

It'd be nice if the code editor wrapped text to the width of the editor pane

I'm not even sure if these things are under our control or not, just writing down my list of minor desires.

**Comment by tcstewar** [comment]
Lots of these are possible: the ace editor is ridiculously extensible.  http://ace.c9.io/

**Comment by jgosmann** [comment]
> It'd be nice if the code editor wrapped text to the width of the editor pane

I would prefer to have this as a setting.

**Comment by bjkomer** [comment]
>>It'd be nice if the code editor wrapped text to the width of the editor pane

>I would prefer to have this as a setting.

I agree, I'd rather have this as a setting. Most of the time I'd prefer the text to not be wrapped.

Another things to add would be tab completion

**Comment by drasmuss** [comment]
Yeah actually thinking about it more I don't think I'd really want text wrapping, especially if we had nice auto-indenting on line breaks.

**Comment by Seanny123** [comment]
Also, ctrl-z undoing the last action while focusing on the editor would be awesome. Then again, I guess the e-macs and vim folks would be frustrated by this? So is the solution just to use your own editor and collapse the system editor? That's still supported right?

**Comment by jgosmann** [comment]
I would like to keep an undo shortcut in the editor. I think, on OS X ctrl+z does exactly what you want and cmd+z is mapped to the editor undo (when the editor has the focus).

**Comment by tcstewar** [comment]
> I would like to keep an undo shortcut in the editor

The ace editor does support doing custom stuff with the undo stack, so it should actually be possible to integrate both the editor actions and the netgraph actions into the same undo system....

**Comment by jgosmann** [comment]
I'm not sure whether I want them in the same system ...

**Comment by tcstewar** [comment]
> I'm not sure whether I want them in the same system ...

I think they should be, because they interact closely.  typing in the editor can create an object that you then move via the mouse.  If I now hit undo twice, I'd expect the first undo to move the object back, and the second undo to remove the typing (which would remove the object).

**Comment by bjkomer** [comment]
> I'm not sure whether I want them in the same system ...

I think I agree with you, but I'm slowly swaying more towards the fence. One of the difficulties of linking them is if someone is using their own editor, we will have no way of undoing their typing. (unless we do something weird like overwriting their file, but if they still have it open in their editor, that could cause issues, and probably not what they want happening)
On the otherhand, so many weird cases can crop up when they are not linked, although I have a feeling many of them will not happen in common use.
